language-support/ts/daml-types/index.ts

```diff
@@ -419,4 +420,82 @@ export const TextMap = <T>(t: Serializable<T>): Serializable<TextMap<T>> => ({
   }
 });
 
-// TODO(MH): `Map` type.
+/**
+ * The counterpart of DAML's `DA.Map.Map K V` type.
+ *
+ * @typeparam K The type of the map keys.
+ * @typeparam V The type of the map values.
+ */
+export interface Map<K, V> {
+  get: (k: K) => V | null;
+  getOr: (k: K, ifNotFound: V) => V;
+  has: (k: K) => boolean;
+  insert: (k: K, v: V) => Map<K, V>;
+  remove: (k: K) => Map<K, V>;
+  keys: () => K[];
+  values: () => V[];
+  kvs: () => [K, V][];
+}
+
+// This code assumes that the decoder is only ever used in decoding values
+// straight from the API responses, and said raw responses are never reused
+// afterwards. This should be enforced by this class not being exported and the
+// daml-ledger module not letting raw JSON responses escape without going
+// through this.
+//
+// Without that assumption, the constructor would need to deep-copy its kvs
+// argument.
+class MapImpl<K, V> implements Map<K, V> {
+  private _kvs: [K, V][];
+  private _keys: K[];
+  private _values: V[];
+  constructor(kvs: [K, V][]) {
+    // sorting done so that generic object deep comparison would find equal
+    // maps equal (as defined by jest's expect().toEqual())
+    this._kvs = _.sortBy(kvs, kv => JSON.stringify(kv[0]));
```

Comment:
I’m a bit surprised lodash doesn’t seem to have a deep ordering helper but I couldn’t find one :disappointed:

Reply:
This was not my first choice either.